### PR TITLE
Add email assertions trait

### PR DESCRIPTION
**Pull request by jadb**
This will allow assertions in tests, like:

``` php
$this->assertEmailTo('Tyler', 'tyler@evilcorp.com');
$this->assertEmailSubject('Tyler, welcome to Evil Corp');
$this->assertEmailTextMessageContains('Hey Tyler!');
```

I haven't written tests yet but will do once everyone thinks the API is good enough.

### CODE REVIEW

File: src/TestSuite/EmailAssertTrait.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+     * @param string $name Sender's name.
+     * @param string $message The failure message to define.
+     * @return void
+     */
+    public function assertEmailTo($email, $name, $message = null)
+    {
+        $expected = [$email => $name];
+        $result = $this->email()->to();
+        $this->assertSame($expected, $result, $message);
+    }
```
Comment (markstory) [inline src/TestSuite/EmailAssertTrait.php]: Should there also be assertions for cc, bcc and attachments?

**Comment by markstory** [comment]
I think this is a neat idea. I like the trait approach more than another base class that requires inheritance.

File: src/TestSuite/EmailAssertTrait.php (hunk 1 above)
Comment (antograssiot) [inline src/TestSuite/EmailAssertTrait.php]: and also maybe a `assertEmailToContains()` to look for an email address in a list

Comment (jadb) [inline src/TestSuite/EmailAssertTrait.php]: Totally, was just a PoC to start with.

**Comment by jadb** [comment]
I have added all suggestions and written some functional tests which would serve as sample. Was not sure how to best unit test it without it just being mocks. Pointers welcomed.

**Comment by markstory** [comment]
Functional tests are more valuable than a pile of mocks would be.

File: src/TestSuite/EmailAssertTrait.php [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        $result = $this->email()->to();
+        $this->assertNotEmpty($result[$email], $message);
+        if ($name !== null) {
+            $this->assertEquals($result[$email], $name, $message);
+        }
+    }
+
+    /**
+     * @param string $filename Expected attachment's filename.
+     * @param array $file Expected attachment's file info.
```
Comment (dereuromark) [inline src/TestSuite/EmailAssertTrait.php]: |null are missing in a few places.

**Comment by ravage84** [comment]
@jadb Just curious. None of your test methods have a short description in the doc block.
I see your method names are quite self descriptive ( :+1:  ), but is this way of doing accepted by our coding standard?

**Comment by jadb** [comment]
@ravage84 no. Just opted not to waste time before I make sure all is done and no methods needs changed or otherwise. But feel free to add descriptions and push to this PR if you want to help :)

File: src/TestSuite/EmailAssertTrait.php [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.1
```
Comment (ADmad) [inline src/TestSuite/EmailAssertTrait.php]: 3.3.3 :)

File: tests/test_app/TestApp/Mailer/TestUserMailer.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * CakePHP(tm) Tests <http://book.cakephp.org/2.0/en/development/testing.html>
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         3.4.0
```
Comment (ADmad) [inline tests/test_app/TestApp/Mailer/TestUserMailer.php]: 3.3.3 :)

**Pull request merged**